res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,719 @@
+    if (loopEnabled > 0) {
+        if (this.loopin4beat[channel]) {
+            newVal = newVal * this.loopAdjustMultiply + engine.getValue(group, 'loop_start_position');
+            engine.setValue(group, 'loop_start_position', newVal);
+            return;
+        }
+        if (this.loopout[channel]) {
+            newVal = newVal * this.loopAdjustMultiply + engine.getValue(group, 'loop_end_position');
+            engine.setValue(group, 'loop_end_position', newVal);
+            return;
+        }
+    }
+
+    if (engine.isScratching(deckNum)) {
+        engine.scratchTick(deckNum, newVal);
+    } else { // fallback
+        engine.setValue(group, 'jog', newVal * this.bendScale);
+    }
+
+};
+
+
+PioneerDDJ400.jogSearch = function(_channel, _control, value, _status, group) {
+    // "highspeed" (scaleup value) pitch bend
+    var newVal = (value - 64) * this.highspeedScale;
+    engine.setValue(group, 'jog', newVal);
+};
+
+PioneerDDJ400.jogTouch = function(channel, _control, value) {
+    var deckNum = channel + 1;
+
+    // skip scratchmode if we adjust the loop points
+    if (this.loopin4beat[channel] || this.loopout[channel]) {
+        return;
+    }
+
+    // on touch jog with vinylmode enabled -> enable scratchmode
+    if (value !== 0 && this.vinylMode) {
+        engine.scratchEnable(deckNum, 720, 33 + 1 / 3, this.alpha, this.beta);
+    } else {
+        // on release jog (value === 0) disable pitch bend mode or scratch mode
+        engine.scratchDisable(deckNum);
+    }
+};
+
+PioneerDDJ400.cycleTempoRange = function(_channel, _control, value, _status, group) {
+    if (value === 0) {
+      return; // ignore release
+    }
+
+    var currRange = engine.getValue(group, 'rateRange');
+    var idx = 0;
+
+    for (var i = 0; i < this.tempoRanges.length; i++) {
+        if (currRange === this.tempoRanges[i]) {
+            idx = (i + 1) % this.tempoRanges.length;
+            break;
+        }
+    }
+    engine.setValue(group, 'rateRange', this.tempoRanges[idx]);
+};
+
+
+function sortAsc(a, b) {
+    return a > b ? 1 : b > a ? -1 : 0;
+}
+
+PioneerDDJ400.initCuePointsAndLoops = function(group) {
+    // create a list of positions in the track which can be selected
+    var points = [];
+
+    for (var padNum = 1; padNum <= 8; padNum++) {
+        points.push(engine.getValue(group, 'hotcue_' + padNum + '_position'));
+    }
+    points.push(engine.getValue(group, 'cue_point'));
+    points.push(engine.getValue(group, 'loop_start_position'));
+    points.push(engine.getValue(group, 'loop_end_position'));
+    points.sort(sortAsc); // sort asc
+
+    return points;
+};
+
+PioneerDDJ400.cueLoopCallLeft = function(_channel, _control, value, _status, group) {
+    if (value === 0) {
+      return; // ignore release
+    }
+
+    var loop_on = engine.getValue(group, 'loop_enabled');
+    if (loop_on) {
+        // loop halve
+        engine.setValue(group, 'loop_scale', 0.5);
```

Comment:
The buttons look like this, so it does make some sense to put 0.5 here.
![400](https://user-images.githubusercontent.com/181628/69528854-56c8ad00-0f6f-11ea-9921-c3b4c87994f4.jpg)

Reply:
It is odd because `loop_scale` does not change `beatloop_size`. However, `beatloop_size` only changes the size of the loop if the loop is exactly the same size, which is very unlikely with a manual loop set without quantize.